### DISCUSSION TITLE
UseFormContext

### DISCUSSION THREAD

**Discussion by HexSleeves** [react-hook-form/react-hook-form#7630] [Q&A]
Curious how most people are setting up their forms in their own apps. Are y’all using prop drilling and passing down register or do you use useFormContext?

seems a bit more performance driven to prop drill since the context causes more rerenders

**Comment by Moshyfawn** [comment]
Depends on how deep you need to prop-drill, how much control you have over components and many more reasons for and against Context vs props. The whole React ecosystem is all about tradeoffs :P
You can take a look at #7611 discussion on how to try to optimize provider pattern. In #7553 discussion there're some points on prop-drilling vs Context vs custom Context. But it sums up with

> as much as I want to put a "best" solution, there is no silver bullet in form application development. It's all case by case.

**Reply by Moshyfawn** [reply]
So, I'd say, do Context if you don't feel like passing props / feels right / stars aligned that way etc. And if you hit any performance issues, see, maybe there's a way to optimize how your form children are getting subscribed to the form state. If there's a serious Context issue _(which there shouldn't be, from my experience)_, consider having a custom "slimmer" Context or just prop-drill

**Reply by HexSleeves** [reply]
Yeah I actually pull out the functions I need from useForm and store them in an atom store from `recoil.js`

now that fieldState api is out I can even limit more rerenders by looking at the form state specifically for a field

**Reply by Moshyfawn** [reply]
Fair enough. I haven't tried taking the form methods above the React tree with Recoil or it's alternatives. It introduces another level of complexity to managing forms and there're quite a few levels here already. But it's just my opinion, which I'm going to try to disperse sometime. And you're right, the `fieldState` API does seem to be complimentary to the atoms pattern
